src/v/archival/async_data_uploader.h

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "base/vassert.h"
+#include "cluster/partition.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/rwlock.hh>
+#include <seastar/core/scheduling.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <boost/iterator/counting_iterator.hpp>
+#include <boost/iterator/transform_iterator.hpp>
+#include <boost/outcome.hpp>
+#include <boost/outcome/outcome.hpp>
+#include <boost/outcome/policy/all_narrow.hpp>
+
+#include <exception>
+#include <utility>
+
+namespace archival {
+
+namespace detail {
+inline model::offset cast_to_offset(int64_t i) noexcept {
+    return model::offset(i);
+}
+} // namespace detail
+
+/// Representation of the inclusive monotonic offset range
+struct inclusive_offset_range {
+    model::offset base;
+    model::offset last;
+
+    inclusive_offset_range() = default;
+
+    inclusive_offset_range(model::offset base, model::offset last)
+      : base(base)
+      , last(last) {
+        vassert(base <= last, "Invalid offset range {} - {}", base, last);
+    }
+
+    bool operator<=>(const inclusive_offset_range&) const = default;
+
+    int64_t size() const noexcept { return last() - base() + 1; }
+
+    bool contains(const model::record_batch_header& hdr) const noexcept {
+        return base <= hdr.base_offset && hdr.last_offset() <= last;
+    }
+
+    auto begin() const noexcept {
+        auto cit = boost::make_counting_iterator<int64_t>(base());
+        return boost::make_transform_iterator(cit, &detail::cast_to_offset);
+    }
+
+    auto end() const noexcept {
+        auto cit = boost::make_counting_iterator<int64_t>(last() + 1);
+        return boost::make_transform_iterator(cit, &detail::cast_to_offset);
+    }
+};
+
+/// Representation of the inclusive monotonic offset range limited by
+/// on-disk size.
+struct size_limited_offset_range {
+    model::offset base;
+    size_t min_size;
+    size_t max_size;
+
+    size_limited_offset_range(
+      model::offset base,
+      size_t max_size,
+      std::optional<size_t> min_size = std::nullopt)
+      : base(base)
+      , min_size(min_size.value_or(0))
+      , max_size(max_size) {}
+
+    bool operator<=>(const size_limited_offset_range&) const = default;
+};
+
+inline std::ostream& operator<<(std::ostream& o, inclusive_offset_range r) {
+    fmt::print(o, "[{}-{}]", r.base, r.last);
+    return o;
+}
+
+/// Result of the upload size calculation.
+/// Contains size of the region in bytes and locks.
```

Comment:
Are the locks yet to be added or stored in some nested structure?

Reply:
they're supposed to be stored while the upload operation is active
they should probably go away once we have new storage layer with MVCC but now it's better to hold them